src/Interpreters/ClusterProxy/executeQuery.cpp

```diff
@@ -117,9 +117,27 @@ void executeQuery(
     else
         throttler = user_level_throttler;
 
+    size_t shards = query_info.cluster->getShardCount();
     for (const auto & shard_info : query_info.cluster->getShardsInfo())
     {
-        stream_factory.createForShard(shard_info, query, query_ast,
+        ASTPtr query_ast_for_shard;
+        if (settings.optimize_skip_unused_shards && settings.optimize_skip_unused_shards_rewrite_in && shards > 1)
```

Comment:
This should take `optimize_skip_unused_shards_limit` into account (i.e. "was skipping done or not"), that is added in #21512

Reply:
Done